network/wsNetwork_test.go

```diff
@@ -1111,6 +1131,866 @@ func TestGetPeers(t *testing.T) {
+	defer netA.Stop()
+	netB.Start()
+	defer netB.Stop()
+
+	addrA, ok := netA.Address()
+	require.True(t, ok)
+	gossipA, err := netA.addrToGossipAddr(addrA)
+	require.NoError(t, err)
+
+	addrB, ok := netB.Address()
+	require.True(t, ok)
+	gossipB, err := netB.addrToGossipAddr(addrB)
+	require.NoError(t, err)
+
+	// set addresses to just host:port to match phonebook/dns format
+	addrA = hostAndPort(addrA)
+	addrB = hostAndPort(addrB)
+
+	// first connection should work just fine
+	if _, ok := netA.tryConnectReserveAddr(addrB); ok {
+		netA.wg.Add(1)
+		netA.tryConnect(addrB, gossipB)
+		// let the tryConnect go forward
+		time.Sleep(100 * time.Millisecond)
+	}
+	// just one A->B connection
+	assert.Equal(t, 0, len(netA.GetPeers(PeersConnectedIn)))
+	assert.Equal(t, 1, len(netA.GetPeers(PeersConnectedOut)))
+	assert.Equal(t, 1, len(netB.GetPeers(PeersConnectedIn)))
+	assert.Equal(t, 0, len(netB.GetPeers(PeersConnectedOut)))
+
+	// confirm identity map was added to for both hosts
+	assert.Equal(t, 1, netA.identityTracker.(*mockIdentityTracker).getSetCount())
+	assert.Equal(t, 1, netA.identityTracker.(*mockIdentityTracker).getInsertCount())
+
+	// netB has to wait for a final verification message over WS Handler, so pause a moment
+	time.Sleep(100 * time.Millisecond)
+	assert.Equal(t, 1, netB.identityTracker.(*mockIdentityTracker).getSetCount())
+	assert.Equal(t, 1, netB.identityTracker.(*mockIdentityTracker).getInsertCount())
+
+	// tell the mock identity tracker that new identies should fail to insert
+	netA.identityTracker.(*mockIdentityTracker).setShouldInsert(false)
+	netB.identityTracker.(*mockIdentityTracker).setShouldInsert(false)
+
+	// bi-directional connection from B should not proceed
+	if _, ok := netB.tryConnectReserveAddr(addrA); ok {
+		netB.wg.Add(1)
+		netB.tryConnect(addrA, gossipA)
+		// let the tryConnect go forward
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	// still just one A->B connection
+	assert.Equal(t, 0, len(netA.GetPeers(PeersConnectedIn)))
+	assert.Equal(t, 1, len(netA.GetPeers(PeersConnectedOut)))
+	assert.Equal(t, 1, len(netB.GetPeers(PeersConnectedIn)))
+	assert.Equal(t, 0, len(netB.GetPeers(PeersConnectedOut)))
+	// netA never attempts to set identity as it never sees a verified identity
+	assert.Equal(t, 1, netA.identityTracker.(*mockIdentityTracker).getSetCount())
+	// netB would attempt to add the identity to the tracker
+	// but it would not end up being added
+	assert.Equal(t, 2, netB.identityTracker.(*mockIdentityTracker).getSetCount())
+	assert.Equal(t, 1, netB.identityTracker.(*mockIdentityTracker).getInsertCount())
+
+	// Check deduplication again, this time from A
+	// the "ok" from tryConnectReserveAddr is overloaded here because isConnectedTo
+	// will prevent this connection from attempting in the first place
+	// in the real world, that isConnectedTo doesn't always trigger, if the hosts are behind
+	// a load balancer or other NAT
+	if _, ok := netA.tryConnectReserveAddr(addrB); ok || true {
```

Comment:
That's right. I wrote it like this for a couple reasons:

* All the tests I wrote that use tryConnect attempt a `tryConnectReserveAddr` because it's how tryConnect gets used in the wsNetwork. In this case I really don't care about the value, but I would rather explicitly overload it and demonstrate/explain the difference than simply not include it.

* At the original time of writing, I was unsure of the potential side effects that `tryConnectReserveAddr` would have on the upcoming `tryConnect`, so I wanted to make sure it matched the real-world code.